File: apps/backend/app/validators.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Iterable, Tuple

from .models import AcceptanceTest, Story
from .schemas import ValidationMessage, ValidationResult

AMBIGUOUS_TERMS = {
    "en": ["optimize", "quickly", "soon", "some", "about", "approximately"],
    "ko": ["빠르게", "적절히", "대략", "어느정도", "최대한"],
}

UNIT_PATTERN = re.compile(r"\\b\\d+(?:\\.\\d+)?\\s?(ms|s|sec|seconds|minutes|px|%|percent|kb|mb)\\b", re.IGNORECASE)
NUMBER_PATTERN = re.compile(r"\\d")
ROLE_PATTERN = re.compile(r"^As a .+", re.IGNORECASE)
GOAL_PATTERN = re.compile(r"^I want .+", re.IGNORECASE)
BENEFIT_PATTERN = re.compile(r"^So that .+", re.IGNORECASE)
GWT_PATTERN = re.compile(r"^Given .+ When .+ Then .+", re.IGNORECASE | re.DOTALL)


@dataclass
class PolicyResult:
    validation: ValidationResult
    blocked: bool


def _collect_ambiguity(text: str, field: str) -> Iterable[ValidationMessage]:
    lowered = text.lower()
    for locale, terms in AMBIGUOUS_TERMS.items():
        for term in terms:
            if term.lower() in lowered:
                yield ValidationMessage(
                    type="ambiguity",
                    message=f"Ambiguous term '{term}' detected",
                    field=field,
                )


def _check_numeric_units(text: str, field: str) -> Iterable[ValidationMessage]:
    if NUMBER_PATTERN.search(text) and not UNIT_PATTERN.search(text):
        yield ValidationMessage(
            type="unit",
            message="Numeric values should include explicit units",
            field=field,
        )


def validate_story(story: Story, *, policy: str = "warn") -> PolicyResult:
    result = ValidationResult()

    if not ROLE_PATTERN.match(story.role.strip()):
        result.errors.append(
            ValidationMessage(type="invest", message="Role must start with 'As a'", field="role")
        )
    if not GOAL_PATTERN.match(story.goal.strip()):
        result.errors.append(
            ValidationMessage(type="invest", message="Goal must start with 'I want'", field="goal")
        )
    if not BENEFIT_PATTERN.match(story.benefit.strip()):
        result.errors.append(
            ValidationMessage(type="invest", message="Benefit must start with 'So that'", field="benefit")
        )

    for field_name in ("title", "role", "goal", "benefit"):
        text = getattr(story, field_name)
        result.warnings.extend(list(_collect_ambiguity(text, field_name)))
        result.warnings.extend(list(_check_numeric_units(text, field_name)))

    blocked = policy == "block" and bool(result.errors)
    return PolicyResult(validation=result, blocked=blocked)


def validate_acceptance_tests(tests: Iterable[AcceptanceTest], *, policy: str = "warn") -> PolicyResult:
    result = ValidationResult()
    for test in tests:
        combined = f"Given {test.given} When {test.when} Then {test.then}"
        if not GWT_PATTERN.match(combined.replace("\n", " ")):
            result.errors.append(
                ValidationMessage(
                    type="gwt",
                    message="Acceptance tests must follow Given/When/Then format",
                    field=test.id,
                )
            )
        for field_name in ("given", "when", "then"):
            text = getattr(test, field_name)
            result.warnings.extend(list(_collect_ambiguity(text, field_name)))
            result.warnings.extend(list(_check_numeric_units(text, field_name)))
    blocked = policy == "block" and bool(result.errors)
    return PolicyResult(validation=result, blocked=blocked)


__all__ = ["validate_story", "validate_acceptance_tests", "AMBIGUOUS_TERMS"]

File: packages/shared/src/validators.ts
import { z } from 'zod';
import { AcceptanceTestSchema, StorySchema } from './schemas';

export const ambiguousDictionary = {
  en: ['optimize', 'quickly', 'soon', 'some', 'about', 'approximately'],
  ko: ['빠르게', '적절히', '대략', '어느정도', '최대한']
};

const numberRegex = /\d/;
const unitRegex = /\b\d+(?:\.\d+)?\s?(ms|s|sec|seconds|minutes|px|%|percent|kb|mb)\b/i;

export type ValidationMessage = {
  type: string;
  field?: string;
  message: string;
};

export type ValidationResult = {
  warnings: ValidationMessage[];
  errors: ValidationMessage[];
};

export const StoryValidator = StorySchema.superRefine((value, ctx) => {
  if (!/^As a .+/i.test(value.role)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['role'],
      message: "Role must start with 'As a'"
    });
  }
  if (!/^I want .+/i.test(value.goal)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['goal'],
      message: "Goal must start with 'I want'"
    });
  }
  if (!/^So that .+/i.test(value.benefit)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      path: ['benefit'],
      message: "Benefit must start with 'So that'"
    });
  }
});

export function analyseAmbiguity(text: string, field: string): ValidationMessage[] {
  const lower = text.toLowerCase();
  const hits: ValidationMessage[] = [];
  for (const list of Object.values(ambiguousDictionary)) {
    for (const term of list) {
      if (lower.includes(term.toLowerCase())) {
        hits.push({ type: 'ambiguity', field, message: `Ambiguous term "${term}" detected` });
      }
    }
  }
  if (numberRegex.test(text) && !unitRegex.test(text)) {
    hits.push({ type: 'unit', field, message: 'Numeric values should include explicit units' });
  }
  return hits;
}

export function validateStory(value: z.infer<typeof StorySchema>): ValidationResult {
  const result = StoryValidator.safeParse(value);
  const warnings = ['title', 'role', 'goal', 'benefit'].flatMap((field) =>
    analyseAmbiguity((value as any)[field], field)
  );
  const errors = [] as ValidationMessage[];
  if (!result.success) {
    errors.push(
      ...result.error.issues.map((issue) => ({
        type: 'invest',
        field: issue.path.join('.'),
        message: issue.message
      }))
    );
  }
  return { warnings, errors };
}

export function validateAcceptanceTests(values: z.infer<typeof AcceptanceTestSchema>[]): ValidationResult {
  const warnings: ValidationMessage[] = [];
  const errors: ValidationMessage[] = [];
  for (const test of values) {
    const combined = `Given ${test.given} When ${test.when} Then ${test.then}`.replace(/\n/g, ' ');
    if (!/^Given .+ When .+ Then .+/i.test(combined)) {
      errors.push({ type: 'gwt', field: test.id, message: 'Acceptance tests must follow Given/When/Then format' });
    }
    ['given', 'when', 'then'].forEach((field) => {
      warnings.push(...analyseAmbiguity((test as any)[field], field));
    });
  }
  return { warnings, errors };
}
